frame/nomination-pools/src/lib.rs

```diff
@@ -0,0 +1,1998 @@
+				ExistenceRequirement::AllowDeath,
+			)
+			.defensive_map_err(|e| e)?;
+			Self::deposit_event(Event::<T>::Withdrawn {
+				delegator: delegator_account.clone(),
+				pool_id: delegator.pool_id,
+				amount: balance_to_unbond,
+			});
+
+			let post_info_weight = if should_remove_pool {
+				ReversePoolIdLookup::<T>::remove(bonded_pool.bonded_account());
+				RewardPools::<T>::remove(delegator.pool_id);
+				Self::deposit_event(Event::<T>::Destroyed { pool_id: delegator.pool_id });
+				SubPoolsStorage::<T>::remove(delegator.pool_id);
+				// Kill accounts from storage by making their balance go below ED. We assume that
+				// the accounts have no references that would prevent destruction once we get to
+				// this point.
+				debug_assert_eq!(
+					T::StakingInterface::total_stake(&bonded_pool.bonded_account())
+						.unwrap_or_default(),
+					Zero::zero()
+				);
+				T::Currency::make_free_balance_be(&bonded_pool.reward_account(), Zero::zero());
+				T::Currency::make_free_balance_be(&bonded_pool.bonded_account(), Zero::zero());
+				bonded_pool.remove();
+				None
+			} else {
+				bonded_pool.dec_delegators().put();
+				SubPoolsStorage::<T>::insert(&delegator.pool_id, sub_pools);
+				Some(T::WeightInfo::withdraw_unbonded_other_update(num_slashing_spans))
+			};
+			Delegators::<T>::remove(&delegator_account);
+
+			Ok(post_info_weight.into())
+		}
+
+		/// Create a new delegation pool.
+		///
+		/// # Arguments
+		///
+		/// * `amount` - The amount of funds to delegate to the pool. This also acts of a sort of
+		///   deposit since the pools creator cannot fully unbond funds until the pool is being
+		///   destroyed.
+		/// * `index` - A disambiguation index for creating the account. Likely only useful when
+		///   creating multiple pools in the same extrinsic.
+		/// * `root` - The account to set as [`PoolRoles::root`].
+		/// * `nominator` - The account to set as the [`PoolRoles::nominator`].
+		/// * `state_toggler` - The account to set as the [`PoolRoles::state_toggler`].
+		///
+		/// # Note
+		///
+		/// In addition to `amount`, the caller will transfer the existential deposit; so the caller
+		/// needs at have at least `amount + existential_deposit` transferrable.
+		#[pallet::weight(T::WeightInfo::create())]
+		#[transactional]
+		pub fn create(
+			origin: OriginFor<T>,
+			amount: BalanceOf<T>,
+			root: T::AccountId,
+			nominator: T::AccountId,
+			state_toggler: T::AccountId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				amount >=
+					T::StakingInterface::minimum_bond()
+						.max(MinCreateBond::<T>::get())
+						.max(MinJoinBond::<T>::get()),
+				Error::<T>::MinimumBondNotMet
+			);
+			ensure!(
+				MaxPools::<T>::get()
+					.map_or(true, |max_pools| BondedPools::<T>::count() < max_pools),
+				Error::<T>::MaxPools
+			);
+			ensure!(!Delegators::<T>::contains_key(&who), Error::<T>::AccountBelongsToOtherPool);
+			ensure!(
+				T::Currency::free_balance(&who) >=
```

Comment:
no need to do this check given the fact that we are transactional -- just transfer out, and `.map_err`

Reply:
although, we could re-write a lot of code if we are to truly leverage transactional... nonetheless this one is also a bit wrong: if you want to check, you should check `ensure_can_withdraw`.

Reply:
d78b16a4a890056b2484bc2b270cae4c26e73197